### PR TITLE
[portinglayer] Matter data model porting

### LINKED ISSUE #76: [portinglayer] matter data model porting
- refer to https://github.com/ambiot/ambz2_matter/pull/143
- do a porting into amebaD
- apart from the makefiles part, the rest of the codes are similar
- you can refer to the examples in ambd_matter to see how the makefiles are designed

### PR DESCRIPTION
**Pull request by SirRu24**
Fix #76 :
- custom data model for ameba series
- able to support dynamic endpoints
- add device type endpoint presets and cluster presets
- light_dm example that uses this ameba data model

### CODE REVIEW

**Comment by step0035** [comment]
- for amebad the `new` and `delete` operator will fail when there is insufficient heap memory in sram, so we overload them to use psram for dynamic memory allocation
- z2 didn't face this issue as it defined both sram and psram as freertos heap region
-  i tried to do this on ameba d but faced some other issues, so using this alternative is easier
- do note that `psram_reserve_malloc` and `psram_reserve_free` might not be thread-safe like `pvPortMalloc` and `vPortFree`, if met with race-condition, you can create a semaphore to lock this heap

**Review by step0035 — APPROVED** [review]

File: component/common/application/matter/example/light_dm/README.md [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    cd ambz2_matter/project/realtek_amebaz2_v0_example/GCC-RELEASE/
+    make is_matter
+    
+### Flash the Image
+Refer to this [guide](https://github.com/ambiot/ambz2_matter/blob/main/tools/AmebaZ2/Image_Tool_Linux/README.md) to flash the image with the Linux Image Tool
+
+### Clean Matter Libraries
+
+    cd ambz2_matter/component/common/application/matter/example/light
+    make clean
```
Comment (xshuqun) [inline component/common/application/matter/example/light_dm/README.md:58]: Please change the build instruction for amebad.

File: project/realtek_amebaD_va0_example/GCC-RELEASE/project_hp/asdk/Makefile [hunk 2]
```diff
@@ -277,6 +279,8 @@ $(LIGHTING_FILE): $(LIGHTING_ZAP)
 	python3 $(CHIPDIR)/scripts/codegen.py --generator cpp-app --output-dir $(CHIPDIR)/examples/lighting-app/ameba/build/chip/codegen --expected-outputs $(MATTER_TOOLDIR)/codegen_helpers/expected.outputs $(CHIPDIR)/examples/lighting-app/lighting-common/lighting-app.matter
 	python3 $(CHIPDIR)/src/app/zap_cluster_list.py --zap_file $(CHIPDIR)/examples/lighting-app/lighting-common/lighting-app.zap > $(CHIPDIR)/examples/lighting-app/ameba/build/chip/codegen/cluster-file.txt
 	python3 $(MATTER_TOOLDIR)/codegen_helpers/parse_clusters.py --cluster_file $(CHIPDIR)/examples/lighting-app/ameba/build/chip/codegen/cluster-file.txt --chip_path $(CHIPDIR)
+	cp $(BASEDIR)/component/common/application/matter/example/light_dm/endpoint_config.h $(CHIPDIR)/examples/lighting-app/ameba/build/chip/codegen/zap-generated
+	cp $(BASEDIR)/component/common/application/matter/example/light_dm/gen_config.h $(CHIPDIR)/examples/lighting-app/ameba/build/chip/codegen/zap-generated
```
Comment (xshuqun) [inline project/realtek_amebaD_va0_example/GCC-RELEASE/project_hp/asdk/Makefile]: Will this affect the build of `make light_porting` and 'make light`?

File: component/common/application/matter/example/light_dm/README.md (hunk 1 above)
Comment (SirRu24) [reply component/common/application/matter/example/light_dm/README.md:58]: Updated

File: project/realtek_amebaD_va0_example/GCC-RELEASE/project_hp/asdk/Makefile (hunk 2 above)
Comment (step0035) [reply project/realtek_amebaD_va0_example/GCC-RELEASE/project_hp/asdk/Makefile:283]: No, the next time you run `make light_port` or `make light`, the `endpoint_config.h` and `gen_config.h` will be re-generated

Comment (step0035) [reply project/realtek_amebaD_va0_example/GCC-RELEASE/project_hp/asdk/Makefile]: yes

Comment (step0035) [reply project/realtek_amebaD_va0_example/GCC-RELEASE/project_hp/asdk/Makefile]: fixed

**Review by step0035 — APPROVED** [review]

**Review by xshuqun — APPROVED** [review]

**Pull request merged**